src/network/networkd-link.c

```diff
@@ -2714,8 +2765,18 @@ int link_update(Link *link, sd_netlink_message *m) {
         }
 
         r = sd_netlink_message_read_strv(m, IFLA_PROP_LIST, IFLA_ALT_IFNAME, &s);
-        if (r >= 0)
+        if (r >= 0) {
+                char **n;
+
+                STRV_FOREACH(n, link->alternative_names)
+                        hashmap_remove_value(link->manager->links_by_name, *n, link);
```

Comment:
if you have that unregistration function i suggested above, you could just call it here, simplifying things quite a bit